### ISSUE TITLE
Make endpoints available in WebSocketContext

### ISSUE THREAD

**Issue by bikeRiver** [unosquare/embedio#291]
I can get IP from
```c#
protected abstract void OnClientConnected(IWebSocketContext context, IPEndPoint localEndPoint, IPEndPoint remoteEndPoint);
```
How to Know IP  from
```c#
protected abstract void OnClientDisconnected(IWebSocketContext context);
```
Thanks.

**Comment by rdeago** [comment]
Hello @bikeRiver, thanks for your input.

From what I see, the only way to have the local and remote endpoints available even after the client disconnects is have them as properties of `IWebSocketContext`.

Since this would require breaking changes, I'm putting this in queue for EmbedIO v3.0.

A quick (and admittedly ugly) workaround could be as follows _(disclaimer: I've not tested it, so it may require additional thinking)_

* Add this code to your `WebSocketsServer`-derived class
```C#
private ConcurrentDictionary<WebSocket, (IPEndPoint, IPEndPoint)> _endPointDict = new ConcurrentDictionary<WebSocket, (IPEndPoint, IPEndPoint)>();
```

* Add the following line at the beginning of your `OnClientConnected` method:
```C#
_endPointDict[context] = (localEndPoint, remoteEndPoint);
```

* Add the following lines at the beginning of your `OnClientDisconnected` method:
```C#
var (localEndPoint, remoteEndPoint) = _endPointDict[context];
_endPointDict.TryRemove(context, out _);
```


---

**Note to assignee:** cache the two endpoints as soon as the context is created, as they could become unavailable later, e.g. if they are requested for the first time after the client disconnects.

**Comment by bikeRiver** [comment]
Thanks. It work

**Comment by rdeago** [comment]
@bikeRiver nice to know that.

I'm reopening the issue, however, because it is now flagged as an enhancement to be implemented in EmbedIO v3.0.

**Comment by rdeago** [comment]
Closed by 54b5d231feab9729feb99f36baa5d484dbcdf389 (although I forgot to mention it in the commit message)
